### PR TITLE
start writing file props on delete instance table

### PR DESCRIPTION
**Pull request by esmadau**
## Description
start writing file props on delete instance table to support external store 

## Related issues
Addresses [[AB#110010](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/110010)].

## Testing
Updated existing tests and added new tests

### CODE REVIEW

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV47.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                            string rSopInstanceUid) = reader.ReadRow(
+                            VLatest.DeletedInstance.Watermark,
+                            VLatest.DeletedInstance.PartitionKey,
+                            VLatest.DeletedInstance.StudyInstanceUid,
+                            VLatest.DeletedInstance.SeriesInstanceUid,
+                            VLatest.DeletedInstance.SopInstanceUid);
+                        _logger.LogInformation(
+                            "Instance queued for deletion. Instance Watermark: {Watermark} , PartitionKey: {PartitionKey}",
+                            rWatermark, rPartitionKey);
+                        _telemetryClient.ForwardLogTrace("Instance queued for deletion", rStudyInstanceUid, rSeriesInstanceUid, rSopInstanceUid);
```
[the remark this answers is not in the document]
Comment (bcarthic) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV47.cs:97]: The data access layer is responsible for interacting with the database and should not be concerned with logging or other concerns unless we have to track any SQL related metrics.  It is easy for maintenance. The service layer, on the other hand, is responsible for coordinating business logic and interacting with other services or components.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Sql/Sprocs/DeleteInstanceV6.sql [hunk 2]
```diff
@@ -36,238 +49,251 @@ BEGIN
...
-    ON DI.InstanceKey = FP.InstanceKey
-
-    -- Deleting tag errors
-    DECLARE @deletedTags AS TABLE
-    (
-        TagKey BIGINT
-    )
-    DELETE XQTE
+        DECLARE @deletedInstances AS TABLE
+                                     (PartitionKey INT,
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Sql/Sprocs/DeleteInstanceV6.sql:53]: Was all these indent intentional?

**Review by wsugarman — CHANGES_REQUESTED** [review]

File: src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs [hunk 3]
```diff
@@ -48,7 +58,8 @@ public class DeleteService : IDeleteService
         EnsureArg.IsNotNull(transactionHandler, nameof(transactionHandler));
         EnsureArg.IsNotNull(logger, nameof(logger));
         EnsureArg.IsNotNull(contextAccessor, nameof(contextAccessor));
-
+        _telemetryClient = EnsureArg.IsNotNull(telemetryClient, nameof(telemetryClient));
+        _isExternalStoreEnabled = EnsureArg.IsNotNull(featureConfiguration?.Value, nameof(featureConfiguration)).EnableExternalStore;
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs:62]: Do we need to 2 null checks?

File: src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs [hunk 4]
```diff
@@ -158,13 +172,24 @@ public async Task<(bool Success, int RetrievedInstanceCount)> CleanupDeletedInst
...
     }
 
+    private void EmitTelemetry(IEnumerable<VersionedInstanceIdentifier> identifiers)
+    {
+        foreach (var identifier in identifiers)
+        {
+            _logger.LogInformation(
+                "Instance queued for deletion. Instance Watermark: {Watermark} , PartitionKey: {PartitionKey} , ExternalStore: {ExternalStore}",
+                identifier.Version, identifier.Partition.Key, _isExternalStoreEnabled);
+            _telemetryClient.ForwardLogTrace("Instance queued for deletion", identifier);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs:182]: Do you think passing the partitionName will be helpful?

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV47.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+using Microsoft.Health.Dicom.Core.Features.Partitioning;
+using Microsoft.Health.Dicom.Core.Models;
+using Microsoft.Health.Dicom.SqlServer.Features.Schema;
+using Microsoft.Health.Dicom.SqlServer.Features.Schema.Model;
+using Microsoft.Health.SqlServer.Features.Client;
+using Microsoft.Health.SqlServer.Features.Storage;
+
+namespace Microsoft.Health.Dicom.SqlServer.Features.Store;
+
+internal class SqlIndexDataStoreV47 : SqlIndexDataStoreV46
```
[the remark this answers is not in the document]
Comment (bcarthic) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV47.cs:24]: +1

File: test/Microsoft.Health.Dicom.Tests.Integration/Persistence/SqlDataStoreTestsFixture.cs [hunk 6]
```diff
@@ -46,6 +49,10 @@ public class SqlDataStoreTestsFixture : IAsyncLifetime
 
     private readonly string _masterConnectionString;
     private readonly SchemaInitializer _schemaInitializer;
+    private readonly TelemetryClient _telemetryClient = new TelemetryClient(new TelemetryConfiguration()
```
Comment (bcarthic) [inline test/Microsoft.Health.Dicom.Tests.Integration/Persistence/SqlDataStoreTestsFixture.cs:52]: I dont think you need this now

**Review by bcarthic — DISMISSED** [review]

File: src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs [hunk 7]
```diff
@@ -172,14 +173,17 @@ public async Task<(bool Success, int RetrievedInstanceCount)> CleanupDeletedInst
         return (success, retrievedInstanceCount);
     }
 
-    private void EmitTelemetry(IEnumerable<VersionedInstanceIdentifier> identifiers)
+    private void EmitTelemetry(IReadOnlyCollection<VersionedInstanceIdentifier> identifiers)
     {
+        _logger.LogInformation(
+            "Instances queued for deletion: {Count}", identifiers.Count);
+        _telemetryClient.ForwardLogTrace($"Instances queued for deletion: {identifiers.Count}");
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs:180]: Perhaps just `$"Queued {identifiers.Count} instance(s) for deletion"`?

File: src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs [hunk 8]
```diff
@@ -172,14 +173,17 @@ public async Task<(bool Success, int RetrievedInstanceCount)> CleanupDeletedInst
         return (success, retrievedInstanceCount);
     }
 
-    private void EmitTelemetry(IEnumerable<VersionedInstanceIdentifier> identifiers)
+    private void EmitTelemetry(IReadOnlyCollection<VersionedInstanceIdentifier> identifiers)
     {
+        _logger.LogInformation(
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Delete/DeleteService.cs:178]: nit: is this newline necessary?

File: src/Microsoft.Health.Dicom.Core/Features/Diagnostic/LogForwarderExtensions.cs [hunk 9]
```diff
@@ -44,37 +45,25 @@ internal static class LogForwarderExtensions
         telemetry.Properties.Add(StudyInstanceUID, instanceIdentifier.StudyInstanceUid);
         telemetry.Properties.Add(SeriesInstanceUID, instanceIdentifier.SeriesInstanceUid);
         telemetry.Properties.Add(SOPInstanceUID, instanceIdentifier.SopInstanceUid);
-        telemetry.Properties.Add(ForwardLogFlag, bool.TrueString);
+        telemetry.Properties.Add(PartitionName, instanceIdentifier.Partition.Name);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Diagnostic/LogForwarderExtensions.cs:48]: Is the forward flag no longer necessary?

File: src/Microsoft.Health.Dicom.Core/Features/Diagnostic/LogForwarderExtensions.cs [hunk 10]
```diff
@@ -44,37 +45,25 @@ internal static class LogForwarderExtensions
...
     /// <summary>
-    /// Emits a trace log with forwarding flag set and adds properties from instanceIdentifier as properties to telemetry.
+    /// Emits a trace log with forwarding flag set.
     /// </summary>
     /// <param name="telemetryClient">client to use to emit the trace</param>
     /// <param name="message">message to set on the trace log</param>
-    /// <param name="StudyInstanceUid">identifier to use to set UIDs on log and telemetry properties</param>
-    /// <param name="SeriesInstanceUid">identifier to use to set UIDs on log and telemetry properties</param>
-    /// <param name="SopInstanceUid">identifier to use to set UIDs on log and telemetry properties</param>
+    /// <remarks>NOTE - do not use this if reporting on any specific instance. Only use as high level remarks. Attempt to use identifiers wherever possible.</remarks>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Diagnostic/LogForwarderExtensions.cs:58]: nit: remarks go under the summary

File: src/Microsoft.Health.Dicom.Core/Features/Diagnostic/LogForwarderExtensions.cs [hunk 11]
```diff
@@ -44,37 +45,25 @@ internal static class LogForwarderExtensions
...
-        EnsureArg.IsNotNull(SeriesInstanceUid, nameof(SeriesInstanceUid));
-        EnsureArg.IsNotNull(SopInstanceUid, nameof(SopInstanceUid));
 
         var telemetry = new TraceTelemetry(message);
-        telemetry.Properties.Add(StudyInstanceUID, StudyInstanceUid);
-        telemetry.Properties.Add(SeriesInstanceUID, SeriesInstanceUid);
-        telemetry.Properties.Add(SOPInstanceUID, SopInstanceUid);
-        telemetry.Properties.Add(ForwardLogFlag, bool.TrueString);
 
         telemetryClient.TrackTrace(telemetry);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Diagnostic/LogForwarderExtensions.cs:68]: Don't we need the forward flag?

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStore.cs [hunk 12]
```diff
@@ -19,7 +19,7 @@
 
 namespace Microsoft.Health.Dicom.SqlServer.Features.Store;
 
-internal sealed class SqlIndexDataStore : IIndexDataStore
+internal class SqlIndexDataStore : IIndexDataStore
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStore.cs:22]: This class should still be `sealed` and not allow inheritance.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV1.cs [hunk 13]
```diff
@@ -83,28 +83,28 @@ public virtual async Task<long> BeginCreateInstanceIndexAsync(Partition partitio
         }
     }
 
-    public virtual async Task<IEnumerable<VersionedInstanceIdentifier>> DeleteInstanceIndexAsync(Partition partition, string studyInstanceUid, string seriesInstanceUid, string sopInstanceUid, DateTimeOffset cleanupAfter, CancellationToken cancellationToken)
+    public virtual Task<IReadOnlyCollection<VersionedInstanceIdentifier>> DeleteInstanceIndexAsync(Partition partition, string studyInstanceUid, string seriesInstanceUid, string sopInstanceUid, DateTimeOffset cleanupAfter, CancellationToken cancellationToken)
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV1.cs:86]: `DeleteInstanceIndexAsync`, `DeleteSeriesIndexAsync`, and `DeleteStudyIndexAsync` do not need to be `virtual`. Only the method `DeleteInstanceAsync`

File: src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV47.cs [hunk 14]
```diff
@@ -29,86 +28,58 @@ public SqlIndexDataStoreV47(SqlConnectionWrapperFactory sqlConnectionWrapperFact
...
-                cleanupAfter,
-                (byte)IndexStatus.Created,
-                partition.Key,
-                studyInstanceUid,
-                seriesInstanceUid,
-                sopInstanceUid);
-
-            try
+            using var reader =
+                await sqlCommandWrapper.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStoreV47.cs:49]: nit: you can put these on the same line

**Review by wsugarman — APPROVED** [review]

**Pull request merged**